addons/mail/static/src/model/model_core.js

```diff
@@ -2,134 +2,304 @@
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {function} patch The function to be applied on the identifying fields.
  */
-function registerInstancePatchModel(modelName, patchName, patch) {
-    _registerPatchModel(modelName, patchName, patch, { type: 'instance' });
+export function patchIdentifyingFields(modelName, patch) {
+    if (!registry.has(modelName)) {
+        throw new Error(`Cannot patch identifyingFields on model "${modelName}": model must be registered before being patched.`);
+    }
+    if (typeof patch !== 'function') {
+        throw new Error(`Cannot patch identifyingFields on model "${modelName}": patch must be a function.`);
+    }
+    patch(registry.get(modelName).get('identifyingFields'));
 }
 
 /**
- * @param {string} name
- * @param {function} factory
- * @param {string[]} [dependencies=[]]
+ * Overrides the lifecycle hooks of the model specified by the `modelName`.
+ *
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {Object} hooks Lifecycle hooks to be overriden.
  */
-function registerNewModel(name, factory, dependencies = []) {
-    const entry = _getEntryFromModelName(name);
-    let entryDependencies = [...dependencies];
-    if (name !== 'mail.model') {
-        entryDependencies = [...new Set(entryDependencies.concat(['mail.model']))];
-    }
-    if (entry.factory) {
-        throw new Error(`Model "${name}" has already been registered!`);
-    }
-    Object.assign(entry, {
-        dependencies: entryDependencies,
-        factory,
-        name,
-    });
+export function patchLifecycleHooks(modelName, hooks) {
+    if (!registry.has(modelName)) {
+        throw new Error(`Cannot patch lifecycle hooks on model "${modelName}": model must be registered before being patched.`);
+    }
+    const definition = registry.get(modelName);
+    for (const [name, handler] of Object.entries(hooks)) {
+        if (typeof handler !== 'function') {
+            throw new Error(`Cannot patch lifecycle hook "${name}" on model "${modelName}": "${name}" must be a function.`);
+        }
+        if (!definition.get('lifecycleHooks').has(name)) {
+            throw new Error(`Cannot patch lifecycle hook "${name}" on model "${modelName}": hook does not exist yet.`);
+        }
+        const hookBeforePatch = definition.get('lifecycleHooks').get(name);
+        definition.get('lifecycleHooks').set(name, function (...args) {
+            this._super = hookBeforePatch;
+            return handler.call(this, ...args);
+        });
+    }
 }
 
-//------------------------------------------------------------------------------
-// Export
-//------------------------------------------------------------------------------
+/**
+ * Overrides the model methods of the model specified by the `modelName`.
+ *
+ * @param {string} modelName The name of the model to which to apply the patch.
+ * @param {Object} methods Model methods to be overriden.
```

Comment:
for all of those object param it would be nice to explain that the key is used as name and the value is used as method/field